### PR TITLE
[PF-1585] Add workspaceByUserFacingId endpoint

### PR DESCRIPTION
**Pull request by melissachang**
Manually tested with swagger (normal case, workspace missing, workspace no permission)

### CODE REVIEW

File: openapi/src/common/parameters.yaml
Comment (jaycarlton) [inline openapi/src/common/parameters.yaml:106]: nit: missing newline at EOF. Really only matters for older POSIX tools.

File: openapi/src/parts/workspace.yaml
Comment (jaycarlton) [inline openapi/src/parts/workspace.yaml:111]: Aside, and maybe a question for @ddietterich: should we draw a distinction between listing workspaces and reading them, or is it our policy that you need read access to know it exists? ISTR it being the latter, which is consistent with my understanding of this new endpoint.

File: service/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java:173]: nit: maybe alphabetize these? There's no natural ordering.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:178]: nit: change `id` to `workspaceUuid` for consistency.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:287]: nit: add `userFacingId` to the log messages, since some UFIDs can look like GUIDs.

File: service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:175]: nit: name `workspaceUuid` for consistency.

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java
Comment (jaycarlton) [inline service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:265]: In the logs, I think it's important to distinguish these, since they won't always look different.

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java
Comment (jaycarlton) [inline service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:365]: nit: never call `get()` without checking/asserting `isPresent()`. Even in test code, it makes me look up to see if you've checked it.

**Review by jaycarlton — APPROVED** [review]
Looks great!

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (ddietterich) [inline service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:273]: This could retrieve the workspace directly in one transaction instead of looping out and back with the UUID to look up the same row again.

File: service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java
Comment (ddietterich) [inline service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:174]: I would prefer to do this operation in one DB transaction instead of two.
You would put the Sam authz call here instead of the `validateWorkspaceAndAction`

If we get a second userFacingId-based API, we will probably want a `validateWorkspaceAndActionByUFID` method.

**Review by ddietterich — CHANGES_REQUESTED** [review]
I would prefer a different structure. What do you think?

File: openapi/src/parts/workspace.yaml
Comment (cbookg) [reply openapi/src/parts/workspace.yaml:111]: Regardless of this API, you will be able to determine this by trying to create a workspace with a given UFID to see if it is rejected.

Comment (cbookg) [reply openapi/src/parts/workspace.yaml:111]: In fact, I was planning to use this API in the "Create Workspace" flow to determine if a provided UFID is taken.

So I was expecting this API to return the UUID, not the full object.  I think before it returns the full object, you would have to have access to use it?

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: If you click and drag, you'll see these are spaces. Intellij has a `Continuation indent` preference that's 8 by default:

image

4 does look better to me, than 8. I changed mine to 4, and I added a step to Terra Onboarding.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: Done, makes sense

File: service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:174]: Done

File: openapi/src/parts/workspace.yaml
Comment (melissachang) [reply openapi/src/parts/workspace.yaml:111]: `getWorkspaceByUserFacingId` will be used for:

- UI: User loads workspace page. Url has userFacingId
- CLI ` terra workspace set --id=userFacingId`

`getWorkspaceByUserFacingId` will only return the workspace if Sam says you have READ access.

For UI determining "is UFID taken", I was thinking that would be a separate endpoint: PF-1605. Something like `/api/workspaces/v1/isUserFacingIdTaken/{workspaceUserFacingId}`. @cbookg do you want separate endpoint, or do you want to call `getWorkspaceByUserFacingId` and 403/404 means ufId is taken?

File: openapi/src/common/parameters.yaml
Comment (ddietterich) [reply openapi/src/common/parameters.yaml:103]: I think this description gets shown to people using the swagger UI, so it is nice to have here for them. Not that we have great descriptions for other things! Just trying to get better incrementally.

File: openapi/src/parts/workspace.yaml
Comment (ddietterich) [reply openapi/src/parts/workspace.yaml:111]: For Jay's question: we only list workspaces that the caller has some access to. We do this by first doing a Sam call to get the list of WORKSPACE objects the caller has access to and then filtering our DAO query with that list.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (ddietterich) [inline service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java:581]: I don't think you need this anymore

**Review by ddietterich — APPROVED** [review]
LGTM - a couple nits

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: Done

File: openapi/src/common/parameters.yaml
Comment (melissachang) [reply openapi/src/common/parameters.yaml]: Ah, I see. Added to all the relevant places, thanks.

**Review by melissachang — COMMENTED** [review]
Will wait for Chris's review before merging

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (jaycarlton) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: If you use the Google Style theme instead of Default, you should get that and other settings for free. I didn't have to change my Continuation indent 
Screen Shot 2022-05-03 at 10 11 36 AM
to 4, for example.

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java
Comment (jaycarlton) [reply service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:365]: Yes it will fail, but not with your assertion failure, so if you have a description in your assertion, you won't see it. One of these libraries has `assertContains(T value, Optional<T> o)` that lets you do it without the get.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (ddietterich) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: Here are the instructions from our DEVELOPER.md describing the standard formatting we are using for MC Terra.

### IntelliJ Setup

1. Open the repo normally (File -> Open)
2. In project structure (the folder icon with a little tetromino over it in the upper
   right corner), make sure the project SDK is set to Java 11. If not, IntelliJ should
   detect it on your system in the dropdown, otherwise click "Add JDK..." and navigate to
   the folder from the last step.
3. Set up [google-java-format](https://github.com/google/google-java-format). We use the
   spotless checker to force code to a standard format. Installing the IntelliJ plug-in
   and library makes it easier to get it in the right format from the start.
4. See some optional tips below in the ["Tips"](#tips) section.

File: openapi/src/parts/workspace.yaml
Comment (cbookg) [reply openapi/src/parts/workspace.yaml:111]: > For UI determining "is UFID taken", I was thinking that would be a separate endpoint: PF-1605. Something like /api/workspaces/v1/isUserFacingIdTaken/{workspaceUserFacingId}. @cbookg do you want separate endpoint, or do you want to call getWorkspaceByUserFacingId and 403/404 means ufId is taken?

Either option is fine with me.  If it is more efficient for the backend to check existing of an ID than to populate the full get result, a distinct ID might be better.

Comment (ddietterich) [reply openapi/src/parts/workspace.yaml:111]: I vote for fewer endpoints. The lookup is a single row in the workspace table and the Sam access check, so returning the full result is not a big cost.

File: service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: I have `google-java-format` plugin enabled. Continuation indent was still 8. I'm guessing [default changed to 8 recently](https://intellij-support.jetbrains.com/hc/en-us/community/posts/4403453685778-Why-is-HTML-continuation-indent-default-8-spaces-), maybe that's why you didn't have to change it @ddietterich .

The [Google scheme](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml) does set continuation indent to 4. Should I add Google Schema to the `IntelliJ Setup` instructions? It's unfortunate that `google-java-format` doesn't take care of everything.

(Side note - There was one instance where even though I used `google-java-format` plugin, `spotlessCheck` still failed. That's why I changed Intellij to run `./gradlew spotlessApply`. I'll let you know next time that happens.)

File: service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java
Comment (melissachang) [reply service/src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:370]: Ah, it makes sense to assert `isPresent()` to get a friendlier error message. Done.

File: openapi/src/parts/workspace.yaml
Comment (melissachang) [reply openapi/src/parts/workspace.yaml:111]: SG. Closed PF-1605 and updated PF-1606.

**Pull request merged**
